File: apina-core/src/main/java/fi/evident/apina/output/ts/AbstractTypeScriptGenerator.java
package fi.evident.apina.output.ts;

import fi.evident.apina.model.*;
import fi.evident.apina.model.parameters.EndpointParameter;
import fi.evident.apina.model.parameters.EndpointPathVariableParameter;
import fi.evident.apina.model.parameters.EndpointRequestParamParameter;
import fi.evident.apina.model.settings.TranslationSettings;
import fi.evident.apina.model.type.ApiArrayType;
import fi.evident.apina.model.type.ApiPrimitiveType;
import fi.evident.apina.model.type.ApiType;
import fi.evident.apina.model.type.ApiTypeName;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static java.lang.String.format;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.joining;

abstract class AbstractTypeScriptGenerator {

    protected final CodeWriter out = new CodeWriter();
    protected ApiDefinition api;
    protected final TranslationSettings settings;
    private final String typePrefix;
    private final String supportPrefix;
    private final String resultFunctor;

    public AbstractTypeScriptGenerator(ApiDefinition api, TranslationSettings settings, String typePrefix, String supportPrefix, String resultFunctor) {
        this.api = requireNonNull(api);
        this.settings = requireNonNull(settings);
        this.typePrefix = requireNonNull(typePrefix);
        this.supportPrefix = requireNonNull(supportPrefix);
        this.resultFunctor = requireNonNull(resultFunctor);
    }

    protected void writeTypes() {

        out.writeExportedInterface("Dictionary<V>", () ->
                out.writeLine("[key: string]: V;"));

        for (ApiTypeName unknownType : api.getAllBlackBoxClasses()) {
            out.writeLine(format("export type %s = {};", unknownType));
        }

        out.writeLine();

        for (EnumDefinition enumDefinition : api.getEnumDefinitions()) {
            out.writeLine(format("export enum %s { %s }", enumDefinition.getType(), String.join(", ", enumDefinition.getConstants())));
        }

        out.writeLine();

        for (ClassDefinition classDefinition : api.getClassDefinitions()) {
            out.writeExportedClass(classDefinition.getType().toString(), () -> {
                for (PropertyDefinition property : classDefinition.getProperties())
                    out.writeLine(property.getName() + ": " + property.getType() + ";");
            });
        }

        writeSerializerDefinitions();
    }

    private String qualifiedTypeName(ApiType type) {
        if (type instanceof ApiPrimitiveType) {
            return type.typeRepresentation();
        } else if (type instanceof ApiArrayType) {
            ApiArrayType arrayType = (ApiArrayType) type;
            return qualifiedTypeName(arrayType.getElementType()) + "[]";
        } else if (settings.isImported(new ApiTypeName(type.typeRepresentation()))) {
            return type.typeRepresentation();
        } else {
            return typePrefix + type.typeRepresentation();
        }
    }

    private void writeSerializerDefinitions() {
        out.write("export function registerDefaultSerializers(config: " + supportPrefix + "ApinaConfig) ").writeBlock(() -> {
            for (ApiTypeName unknownType : api.getAllBlackBoxClasses()) {
                out.write("config.registerIdentitySerializer(").writeValue(unknownType.toString()).writeLine(");");
            }
            out.writeLine();

            for (EnumDefinition enumDefinition : api.getEnumDefinitions()) {
                String enumName = enumDefinition.getType().toString();
                out.write("config.registerEnumSerializer(").writeValue(enumName).write(", ");
                out.write(enumName).writeLine(");");
            }
            out.writeLine();

            for (ClassDefinition classDefinition : api.getClassDefinitions()) {
                Map<String, String> defs = new LinkedHashMap<>();

                for (PropertyDefinition property : classDefinition.getProperties())
                    defs.put(property.getName(), typeDescriptor(property.getType()));

                out.write("config.registerClassSerializer(").writeValue(classDefinition.getType().toString()).write(", ");
                out.writeValue(defs).writeLine(");");
                out.writeLine();
            }
        });

        out.writeLine().writeLine();
    }

    protected static Map<String, Object> createConfig(Endpoint endpoint) {
        Map<String, Object> config = new LinkedHashMap<>();

        config.put("uriTemplate", endpoint.getUriTemplate().toString());
        config.put("method", endpoint.getMethod().toString());

        List<EndpointPathVariableParameter> pathVariables = endpoint.getPathVariables();
        if (!pathVariables.isEmpty())
            config.put("pathVariables", createPathVariablesMap(pathVariables));

        List<EndpointRequestParamParameter> requestParameters = endpoint.getRequestParameters();
        if (!requestParameters.isEmpty())
            config.put("requestParams", createRequestParamMap(requestParameters));

        endpoint.getRequestBody().ifPresent(body -> config.put("requestBody", serialize(body.getName(), body.getType().unwrapNullable())));
        endpoint.getResponseBody().ifPresent(body -> config.put("responseType", typeDescriptor(body)));

        return config;
    }

    protected String endpointSignature(Endpoint endpoint) {
        String name = endpoint.getName();
        String parameters = parameterListCode(endpoint.getParameters());
        String resultType = endpoint.getResponseBody().map(this::qualifiedTypeName).orElse("void");

        return format("%s(%s): %s<%s>", name, parameters, resultFunctor, resultType);
    }

    private String parameterListCode(List<EndpointParameter> parameters) {
        return parameters.stream()
                .map(p -> p.getName() + ": " + qualifiedTypeName(p.getType()))
                .collect(joining(", "));
    }

    private static Map<String, Object> createRequestParamMap(Collection<EndpointRequestParamParameter> parameters) {
        Map<String, Object> result = new LinkedHashMap<>();

        for (EndpointRequestParamParameter param : parameters)
            result.put(param.getQueryParameter(), serialize(param.getName(), param.getType()));

        return result;
    }

    private static Map<String, Object> createPathVariablesMap(List<EndpointPathVariableParameter> pathVariables) {
        Map<String, Object> result = new LinkedHashMap<>();

        for (EndpointPathVariableParameter param : pathVariables)
            result.put(param.getPathVariable(), serialize(param.getName(), param.getType()));

        return result;
    }

    /**
     * Returns TypeScript code to serialize {@code variable} of given {@code type}
     * to transfer representation.
     */
    private static RawCode serialize(String variable, ApiType type) {
        return new RawCode("this.context.serialize(" + variable + ", '" + typeDescriptor(type) + "')");
    }

    private static String typeDescriptor(ApiType type) {
        // Use ApiType's native representation as type descriptor.
        // This method encapsulates the call to make it meaningful in this context.
        return type.unwrapNullable().typeRepresentation();
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/output/ts/AbstractTypeScriptGenerator.kt
package fi.evident.apina.output.ts

import fi.evident.apina.model.*
import fi.evident.apina.model.parameters.EndpointParameter
import fi.evident.apina.model.settings.EnumMode
import fi.evident.apina.model.settings.OptionalTypeMode
import fi.evident.apina.model.settings.TranslationSettings
import fi.evident.apina.model.settings.TypeWriteMode
import fi.evident.apina.model.type.ApiType
import fi.evident.apina.model.type.ApiTypeName
import fi.evident.apina.output.common.RawCode
import fi.evident.apina.utils.readResourceAsString
import java.lang.String.format

abstract class AbstractTypeScriptGenerator(
    val api: ApiDefinition,
    val settings: TranslationSettings,
    private val resultFunctor: ResultFunctor,
    private val classDecorator: String,
    private val platformRuntimeCodePath: String,
    private val platformSpecificImports: Map<String, List<String>> = emptyMap()
) {

    private val out = TypeScriptWriter()

    val output: String
        get() = out.output

    fun writeApi() {
        writeHeader()
        writeImports()
        writeTypes()
        writeResource("typescript/runtime-common.ts")
        writeResource(platformRuntimeCodePath)
        writeEndpoints(api.endpointGroups)
        writeApinaConfig()
    }

    private fun writeHeader() {
        // The project using our generated code might generate the code to directory where
        // it's checked by TSLint and we don't know what settings are used so just disable
        // TSLint for the whole file.
        out.writeLine("/* tslint:disable */")
    }

    private fun writeImports() {
        val imports = settings.imports

        if (imports.isNotEmpty()) {
            for (anImport in imports)
                out.writeImport(anImport.moduleName, anImport.types.map { it.name })

            out.writeLine()
        }

        if (platformSpecificImports.isNotEmpty()) {
            for ((module, imps) in platformSpecificImports)
                out.writeImport(module, imps)

            out.writeLine()
        }

        if (settings.reexportImports && imports.isNotEmpty()) {
            for (anImport in imports)
                out.writeExport(anImport.types.map { it.name })

            out.writeLine()
        }
    }

    private fun writeTypes() {

        for (type in settings.brandedPrimitiveTypes)
            out.writeLine("export type ${type.brandedType.name} = Branded<${type.implementationType.toTypeScript()}, '${type.brandedType.name}'>;")

        for (type in api.allBlackBoxClasses)
            out.writeLine("export type ${type.name} = {};")

        for ((alias, target) in api.typeAliases)
            out.writeLine("export type ${alias.name} = ${target.toTypeScript()};")

        out.writeLine()

        for (enumDefinition in api.enumDefinitions)
            writeEnum(enumDefinition)

        out.writeLine()

        val classDefinitionWriter = when (settings.typeWriteMode) {
            TypeWriteMode.CLASS -> out::writeExportedClass
            TypeWriteMode.INTERFACE -> out::writeExportedInterface
        }

        for (classDefinition in api.structuralTypeDefinitions) {
            classDefinitionWriter(classDefinition.type.name) {
                for (property in classDefinition.properties)
                    if (settings.optionalTypeMode == OptionalTypeMode.UNDEFINED && property.type is ApiType.Nullable) {
                        out.writeLine("${property.name}?: ${property.type.type.toTypeScript()};")
                    } else {
                        out.writeLine("${property.name}: ${property.type.toTypeScript()};")
                    }
            }
        }

        for (definition in api.discriminatedUnionDefinitions)
            writeDiscriminatedUnion(definition)
    }

    private fun writeEnum(enumDefinition: EnumDefinition) {
        val typeName = enumDefinition.type.name
        val constants = enumDefinition.constants
        out.writeLine(
            when (settings.enumMode) {
                EnumMode.DEFAULT ->
                    format("export enum %s { %s }", typeName, constants.joinToString(", ") { "$it = \"$it\"" })

                EnumMode.STRING_UNION ->
                    format("export type %s = %s;", typeName, constants.joinToString(" | ") { "\"$it\"" })

                EnumMode.INT_ENUM ->
                    format("export enum %s { %s }", typeName, constants.joinToString(", "))
            }
        )
    }

    private fun writeDiscriminatedUnion(definition: DiscriminatedUnionDefinition) {
        // First individual members of the union...
        for ((discriminatorValue, member) in definition.types) {
            val typeName = discriminatedUnionMemberType(definition.type, member)
            out.writeBlock("export interface $typeName extends ${member.type.name}") {
                out.writeLine("${definition.discriminator}: '$discriminatorValue';")
            }
            out.writeLine()
        }

        // ...then the union itself
        val members = definition.types.values.map { discriminatedUnionMemberType(definition.type, it) }
        out.writeLine("export type ${definition.type.name} = ${members.joinToString(" | ")};")
        out.writeLine()
    }

    private fun writeEnumSerializer(enumDefinition: EnumDefinition) {
        when (settings.enumMode) {
            EnumMode.INT_ENUM -> {
                val enumName = enumDefinition.type.name
                out.write("this.registerEnumSerializer(").writeValue(enumName).write(", ").write(enumName)
                    .writeLine(");")
            }

            EnumMode.DEFAULT, EnumMode.STRING_UNION ->
                writeIdentitySerializer(enumDefinition.type)
        }
    }

    private fun writeIdentitySerializer(type: ApiTypeName) {
        out.write("this.registerIdentitySerializer(").writeValue(type.name).writeLine(");")
    }

    private fun writeApinaConfig() {
        out.writeBlock("export class ApinaConfig extends ApinaConfigBase") {
            out.writeBlock("constructor()") {
                out.writeLine("super();")
                out.writeLine()

                writeSerializerDefinitions()
            }
        }
    }

    private fun writeSerializerDefinitions() {
        for (type in settings.brandedPrimitiveTypes)
            writeIdentitySerializer(type.brandedType)

        for (aliasedType in api.typeAliases.keys)
            writeIdentitySerializer(aliasedType)

        for (unknownType in api.allBlackBoxClasses)
            writeIdentitySerializer(unknownType)

        out.writeLine()

        for (enumDefinition in api.enumDefinitions)
            writeEnumSerializer(enumDefinition)

        out.writeLine()

        for (classDefinition in api.structuralTypeDefinitions) {
            val defs = LinkedHashMap<String, String>()

            for (property in classDefinition.properties)
                defs[property.name] = typeDescriptor(property.type)

            val typeName = classDefinition.type.name
            out.write("this.registerClassSerializer<$typeName>(").writeValue(typeName).write(", ")
            out.writeValue(defs).writeLine(");")
            out.writeLine()
        }

        for (definition in api.discriminatedUnionDefinitions) {
            val defs = mutableMapOf<String, String>()

            for ((discriminatorValue, type) in definition.types)
                defs[discriminatorValue] = typeDescriptor(ApiType.Class(type.type))

            val typeName = definition.type.name
            out.write("this.registerDiscriminatedUnionSerializer<$typeName>(")
            out.writeValue(typeName).write(", ")
            out.writeValue(definition.discriminator).write(", ")
            out.writeValue(defs).writeLine(");")
            out.writeLine()
        }
    }

    private fun writeResource(path: String) {
        out.write(readResourceAsString(path))
        out.writeLine()
    }

    private fun writeEndpoints(endpointGroups: Collection<EndpointGroup>) {
        for (endpointGroup in endpointGroups) {
            if (classDecorator.isNotBlank()) {
                out.writeLine(classDecorator)
            }
            out.writeBlock("export class ${endpointGroup.name}Endpoint") {

                out.writeBlock("constructor(private readonly context: ApinaEndpointContext)") { }

                for (endpoint in endpointGroup.endpoints) {
                    writeEndpoint(endpoint)
                    out.writeLine().writeLine()

                    if (endpoint.generateUrlMethod) {
                        writeUrlEndpoint(endpoint)
                        out.writeLine().writeLine()
                    }
                }
            }
        }
    }

    private fun writeEndpoint(endpoint: Endpoint) {
        val parameters = parameterListCode(endpoint.parameters)
        val resultType = endpoint.responseBody?.let { qualifiedTypeName(it) } ?: "void"
        val signature = "${endpoint.name}($parameters): ${resultFunctor.apply(resultType)}"

        out.write(signature).write(" ").writeBlock {
            out.write("return this.context.request(")
                .writeValue(createConfig(endpoint)).writeLine(");")
        }
    }

    private fun writeUrlEndpoint(endpoint: Endpoint) {
        val parameters = parameterListCode(endpoint.urlParameters)
        val signature = "${endpoint.name}Url($parameters): string"

        out.write(signature).write(" ").writeBlock {
            out.write("return this.context.url(")
                .writeValue(createConfig(endpoint, onlyUrl = true))
                .writeLine(");")
        }
    }

    private fun qualifiedTypeName(type: ApiType): String = when {
        type is ApiType.Nullable -> when (settings.optionalTypeMode) {
            OptionalTypeMode.UNDEFINED -> qualifiedTypeName(type.type) + " | undefined"
            OptionalTypeMode.NULL -> qualifiedTypeName(type.type) + " | null"
        }

        type is ApiType.Primitive -> type.toTypeScript()
        type is ApiType.Array -> qualifiedTypeName(type.elementType) + "[]"
        settings.isImportedOrBrandedType(ApiTypeName(type.toTypeScript())) -> type.toTypeScript()
        else -> type.toTypeScript()
    }

    private fun discriminatedUnionMemberType(unionType: ApiTypeName, memberType: ClassDefinition) =
        "${unionType.name}_${memberType.type.name}"

    private fun parameterListCode(parameters: List<EndpointParameter>): String {
        val firstOptionalIndex = parameters.indexOfLast { it.type !is ApiType.Nullable } + 1

        return parameters.withIndex().joinToString(", ") { (index, p) ->
            if (p.type is ApiType.Nullable) {
                val type = qualifiedTypeName(p.type.type)
                if (index >= firstOptionalIndex)
                    "${p.name}?: $type | null"
                else
                    "${p.name}: $type | null | undefined"
            } else {
                "${p.name}: ${qualifiedTypeName(p.type)}"
            }
        }
    }

    private fun createConfig(endpoint: Endpoint, onlyUrl: Boolean = false): Map<String, Any> = buildMap {
        put("uriTemplate", endpoint.uriTemplate.toString())

        if (!onlyUrl)
            put("method", endpoint.method.toString())

        val pathVariables = endpoint.pathVariables
        if (pathVariables.isNotEmpty())
            put("pathVariables", pathVariables.associate { it.pathVariable to serialize(it.name, it.type) })

        val requestParameters = endpoint.requestParameters
        if (requestParameters.isNotEmpty())
            put("requestParams", requestParameters.associate { it.queryParameter to serialize(it.name, it.type) })

        if (!onlyUrl) {
            endpoint.requestBody?.let { body ->
                put("requestBody", serialize(body.name, body.type.unwrapNullable()))
            }

            endpoint.responseBody?.let { body ->
                put("responseType", typeDescriptor(body))
            }
        }
    }

    /**
     * Returns TypeScript code to serialize `variable` of given `type`
     * to transfer representation.
     */
    private fun serialize(variable: String, type: ApiType) =
        RawCode("this.context.serialize(" + variable + ", '" + typeDescriptor(type) + "')")

    private fun typeDescriptor(type: ApiType): String {
        // Use ApiType's native representation as type descriptor.
        // This method encapsulates the call to make it meaningful in this context.
        return type.unwrapNullable().toTypeScript()
    }

    private fun ApiType.toTypeScript(): String = toTypeScript(settings.optionalTypeMode)

    companion object {

        /**
         * When writing TypeScript, members of discriminated unions will produce
         * interface definitions, so we'll include them as logical class definitions
         * even though they are not in the model.
         */
        private val ApiDefinition.structuralTypeDefinitions: Collection<ClassDefinition>
            get() = classDefinitions + discriminatedUnionDefinitions
                .flatMap { du -> du.types.values }
                .distinctBy { it.type }
                .sortedBy { it.type }
    }
}

internal fun ApiType.toTypeScript(optionalTypeMode: OptionalTypeMode): String = when (this) {
    is ApiType.Array -> elementType.toTypeScript(optionalTypeMode) + "[]"
    is ApiType.BlackBox -> name.name
    is ApiType.Class -> name.name
    is ApiType.Dictionary -> "Record<string, ${valueType.toTypeScript(optionalTypeMode)}>"
    is ApiType.Nullable -> when (optionalTypeMode) {
        OptionalTypeMode.UNDEFINED -> type.toTypeScript(optionalTypeMode) + " | undefined"
        OptionalTypeMode.NULL -> type.toTypeScript(optionalTypeMode) + " | null"
    }

    is ApiType.Primitive -> toTypeScript()
}

private fun ApiType.Primitive.toTypeScript(): String = when (this) {
    ApiType.Primitive.ANY -> "any"
    ApiType.Primitive.STRING -> "string"
    ApiType.Primitive.BOOLEAN -> "boolean"
    ApiType.Primitive.INTEGER -> "number"
    ApiType.Primitive.FLOAT -> "number"
    ApiType.Primitive.VOID -> "void"
}
